src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -79,62 +79,87 @@ public T runWithAnyNode() {
+    ConnectionGetter redirected = null;
+
+    for (int currentAttempt = 0; currentAttempt < this.maxAttempts; currentAttempt++) {
+      Jedis connection = null;
+      try {
+        if (redirected != null) {
+          connection = redirected.getConnection();
         } else {
-          connection = connectionHandler.getConnectionFromSlot(slot);
+          connection = connectionGetter.getConnection();
+        }
+        return execute(connection);
+      } catch (JedisNoReachableClusterNodeException e) {
+        throw e;
+      } catch (JedisConnectionException e) {
+        connectionGetter = handleConnectionProblem(slot, currentAttempt);
+      } catch (JedisRedirectionException e) {
+        redirected = handleRedirection(connection, e);
+      } finally {
+        if (connection != null) {
+          releaseConnection(connection);
         }
       }
+    }
 
-      return execute(connection);
+    throw new JedisClusterMaxAttemptsException("No more cluster attempts left.");
+  }
 
-    } catch (JedisNoReachableClusterNodeException jnrcne) {
-      throw jnrcne;
-    } catch (JedisConnectionException jce) {
-      // release current connection before recursion
-      releaseConnection(connection);
-      connection = null;
-
-      if (attempts <= 1) {
-        //We need this because if node is not reachable anymore - we need to finally initiate slots
-        //renewing, or we can stuck with cluster state without one node in opposite case.
-        //But now if maxAttempts = [1 or 2] we will do it too often.
-        //TODO make tracking of successful/unsuccessful operations for node - do renewing only
-        //if there were no successful responses from this node last few seconds
-        this.connectionHandler.renewSlotCache();
-      }
+  private ConnectionGetter handleConnectionProblem(final int slot, int currentAttempt) {
+    int attemptsLeft = (maxAttempts - currentAttempt) - 1;
+    if (attemptsLeft <= 1) {
+      //We need this because if node is not reachable anymore - we need to finally initiate slots
+      //renewing, or we can stuck with cluster state without one node in opposite case.
+      //But now if maxAttempts = [1 or 2] we will do it too often.
+      //TODO make tracking of successful/unsuccessful operations for node - do renewing only
+      //if there were no successful responses from this node last few seconds
+      this.connectionHandler.renewSlotCache();
+    }
 
-      return runWithRetries(slot, attempts - 1, tryRandomNode, redirect);
-    } catch (JedisRedirectionException jre) {
-      // if MOVED redirection occurred,
-      if (jre instanceof JedisMovedDataException) {
-        // it rebuilds cluster's slot cache recommended by Redis cluster specification
-        this.connectionHandler.renewSlotCache(connection);
+    return new ConnectionGetter() {
+      @Override
+      public Jedis getConnection() {
+        return connectionHandler.getConnectionFromSlot(slot);
       }
+    };
```

Comment:
Repeated code (ii).

Reply:
See above, `slot` can be different between these two so they just *look* identical.